### PR TITLE
better handle missing I2C muxes

### PR DESCRIPTION
**Pull request by bcantrill**
Draft commit comment:

```
After the fix to #1413, the system is unwilling to ignore mux related
errors -- which has generated a problem on some lab systems that are
missing muxes:  because we can never get the bus into a known mux
state these systems now become entirely unusable, as even devices that
aren't on the missing segments are not accessible.  This fixes that by
observing that a mux that is affirmatively missing -- that is, one
that doesn't reply to its in-band management at all -- can be assumed
to have segments that are similarly missing (and therefore as good as
disabled).  This (naturally) doesn't change the fact that accessing
any devices attached to the missing mux will generate an error, but it
allows the system to broadly drive on absent accesses to those
devices.
```

### CODE REVIEW

File: drv/stm32xx-i2c-server/src/main.rs [hunk 1]
```diff
@@ -149,13 +149,38 @@ fn configure_mux(
...
+            // need to successfully talk to every mux (or successfully learn
+            // that the mux is gone entirely!), and disable every segment.  If
+            // there is any failure through here that isn't the mux being
+            // affirmatively gone, we'll just return the error, leaving our
+            // mux state as unknown.
             //
             all_muxes(controller, port, muxes, |mux| {
-                mux.driver.enable_segment(mux, controller, None, ctrl)
+                match mux.driver.enable_segment(mux, controller, None, ctrl) {
+                    Err(code) if code == ResponseCode::MuxMissing => {
```
Comment (cbiffle) [inline drv/stm32xx-i2c-server/src/main.rs:161]: Nit:
```suggestion
                    Err(ResponseCode::MuxMissing) => {
```

File: drv/stm32xx-i2c-server/src/main.rs [hunk 2]
```diff
@@ -149,13 +149,38 @@ fn configure_mux(
...
+                        // the bus.  So on this error (and only this error), we
+                        // note that we saw it, and drive on.  (Note that
+                        // attempting to speak to a device on a segment on the
+                        // missing mux will properly return MuxMissing -- and
+                        // set our bus's mux state to be unknown.)
+                        //
+                        ringbuf_entry!(Trace::MuxMissing(mux.address));
+                        Ok(())
+                    }
+                    Err(code) => Err(code),
```
Comment (cbiffle) [inline drv/stm32xx-i2c-server/src/main.rs:181]: I think these lines are equivalent to `other => other`, fwiw.

**Review by cbiffle — COMMENTED** [review]
Huh, this diff is way smaller than I was expecting. So the key is just permitting a single error condition on the `enable_segment` path?

**Comment by bcantrill** [comment]
> Huh, this diff is way smaller than I was expecting. So the key is just permitting a single error condition on the `enable_segment` path?

Yeah -- the other paths actually handle themselves, it turns out.  When we treat a missing mux as implying that all of its segments are in fact disabled, everything else just... works.

File: drv/stm32xx-i2c-server/src/main.rs (hunk 2 above)
Comment (bcantrill) [reply drv/stm32xx-i2c-server/src/main.rs:181]: 🤯🤯🤯  **HOW LONG HAS THAT BEEN THERE?!**

Comment (bcantrill) [reply drv/stm32xx-i2c-server/src/main.rs:181]: That is a *really* useful pattern (ha!) that I basically assumed (for terrible reasons?) didn't/wouldn't work.

**Review by cbiffle — APPROVED** [review]

**Pull request merged**